### PR TITLE
OCF RA: Add vhost parameter to rabbitmq-server-ha.ocf

### PR DESCRIPTION
**Pull request by vuntz**
This enables the cluster to focus on a vhost that is not /, in case the
most important vhost is something else.

For reference, other vhosts may exist in the cluster, but these are not
guaranteed to not suffer from any data loss. This patch doesn't address
this issue.

Closes https://github.com/rabbitmq/rabbitmq-server-release/issues/22

### CODE REVIEW

**Comment by michaelklishin** [comment]
@binarin one more for ya ;)

File: scripts/rabbitmq-server-ha.ocf [hunk 1]
```diff
@@ -66,6 +67,7 @@ OCF_RESKEY_rmq_feature_local_list_queues_default=true
 : ${OCF_RESKEY_mnesia_schema_base=${OCF_RESKEY_mnesia_schema_base_default}}
 : ${OCF_RESKEY_pid_file=${OCF_RESKEY_pid_file_default}}
 : ${OCF_RESKEY_node_port=${OCF_RESKEY_node_port_default}}
+: ${OCF_RESKEY_vhost=${OCF_RESKEY_vhost_default}}
```
Comment (binarin) [inline scripts/rabbitmq-server-ha.ocf]: I suggest renaming it to `OCF_RESKEY_default_vhost` ( and `OCF_RESKEY_default_vhost_default` ) to match the config option name in `rabbitmq.config` . And mentioning somewhere in comments that using this option is mandatory for proper functioning of the script.

Comment (vuntz) [reply scripts/rabbitmq-server-ha.ocf]: The option is renamed as suggested; I also extended the documentation for this option to stress out that it must be configured correctly.

File: scripts/rabbitmq-server-ha.ocf [hunk 2]
```diff
@@ -260,6 +262,16 @@ ${OCF_RESKEY_binary} should listen on this port
 <content type="boolean" default="${OCF_RESKEY_node_port_default}" />
 </parameter>
 
+<parameter name="vhost" unique="0" required="0">
+<longdesc lang="en">
+Main virtual host used in the cluster, guaranteed to not suffer data loss. Other
```
Comment (michaelklishin) [inline scripts/rabbitmq-server-ha.ocf]: This is a bold claim. Under what conditions? What or who offers the guarantee?

There is no "main" or "special" vhost in RabbitMQ. Please reword this to be less confusing.

File: scripts/rabbitmq-server-ha.ocf [hunk 3]
```diff
@@ -260,6 +262,16 @@ ${OCF_RESKEY_binary} should listen on this port
...
+virtual hosts may exist, but some scenarios can lead to data loss for them. This
```
Comment (michaelklishin) [inline scripts/rabbitmq-server-ha.ocf]: This provides no details but sounds really alarming. Please reword this.

File: scripts/rabbitmq-server-ha.ocf [hunk 4]
```diff
@@ -260,6 +262,16 @@ ${OCF_RESKEY_binary} should listen on this port
...
+must be correctly configured for the cluster to behave as intended.
+</longdesc>
+<shortdesc lang="en">Main virtual host</shortdesc>
```
Comment (michaelklishin) [inline scripts/rabbitmq-server-ha.ocf]: There is no "main" or "special" vhost in RabbitMQ. Please reword this to be less confusing.

**Review by michaelklishin — CHANGES_REQUESTED** [review]

**Comment by michaelklishin** [comment]
Thank you. I think the wording of the docs cannot be merged as is. It is imprecise, confusing and alarming.

File: scripts/rabbitmq-server-ha.ocf (hunk 4 above)
Comment (michaelklishin) [reply scripts/rabbitmq-server-ha.ocf]: In case you are looking for naming recommendations, I'd use "default vhost" — whiles still not very precise — instead of "main vhost."

File: scripts/rabbitmq-server-ha.ocf (hunk 3 above)
Comment (michaelklishin) [reply scripts/rabbitmq-server-ha.ocf]: What sounds really confusing to me is that all vhosts in RabbitMQ are stored more or less next to each other. Without knowing much what the modern version of this file does, I cannot think of why it would work this way. And I am not a beginner.

**Comment by vuntz** [comment]
@michaelklishin see https://github.com/rabbitmq/rabbitmq-server-release/issues/22#issuecomment-290342118 for the background. I'll sit a bit on it as I can't find a much better phrasing for this; if anyone has suggestions for this, that's welcome, else I'll come back to it later today or tomorrow with hopefully some improvement.

File: scripts/rabbitmq-server-ha.ocf (hunk 3 above)
Comment (binarin) [reply scripts/rabbitmq-server-ha.ocf]: I think that you're putting too much hope in what this script can do. It just strives to minimize the data loss by giving a broker some time to finish queue synchronization before taking some destructive actions. But there is no guarantees at all that data won't be lost - i.e. in some worst case scenario can lose all the data. And in this worst case scenario you'll **lose** not only all the messages, but also **all custom vhosts** themselves - only `default_vhost` from `rabbitmq.config`  will exist afterwards.

This is the price that you are paying for the fact that OCF script will **automatically** recover cluster to a working state from any starting state.

**Comment by michaelklishin** [comment]
@vuntz sure. If you and @binarin can explain what the possible scenarios are for this script, I'd be happy to help with that. However, we cannot make promises about what data will or will not be retained in general as there are factors that the script cannot possibly account for.

**Comment by vuntz** [comment]
I've completely rephrased the doc part to explain that the vhost is used for checking the synchronization status. Is this better?

**Comment by michaelklishin** [comment]
@vuntz definitely, thank you!

**Review by michaelklishin — APPROVED** [review]

**Comment by michaelklishin** [comment]
@vuntz @binarin should this (and other recent contributions from @vuntz) be backported to `3.6.x`/`stable`?

**Comment by vuntz** [comment]
@michaelklishin thanks for the quick action!

About the backport: tbh, I'm not familiar with the backporting policy for rabbitmq, so I can't properly answer, but my non-informed answer would be yes ;-)

**Comment by michaelklishin** [comment]
well, if your build system picks this file from RabbitMQ release artifacts and you want to get this in the next 3.6.x release, it must be backported. If you fetch it from GitHub in master, then backporting would change nothing for you.

**Comment by vuntz** [comment]
@michaelklishin for the package in the openSUSE build service, the decision won't matter as I'll add the patch to the package before the next release. So the question is more about whether it's useful or not for others. I would think so; but maybe it's no big deal for now.

**Comment by michaelklishin** [comment]
unless @binarin suggests otherwise I think we can cherry-pick the contributions
